src/server/tiered_storage.cc

```diff
@@ -346,53 +341,22 @@ void TieredStorage::SetMemoryLowLimit(size_t mem_limit) {
 
 util::fb2::Future<string> TieredStorage::Read(DbIndex dbid, string_view key,
                                               const PrimeValue& value) {
-  DCHECK(value.IsExternal());
-  util::fb2::Future<string> future;
-  if (value.IsCool()) {
-    // If we read a cool record - bring it back to primary table.
-    PrimeValue hot = Warmup(dbid, value.GetCool());
-    string tmp;
-
-    DCHECK_EQ(value.Size(), hot.Size());
-    hot.GetString(&tmp);
-    future.Resolve(tmp);
-
-    // TODO: An awful hack - to fix later.
-    const_cast<PrimeValue&>(value) = std::move(hot);
-  } else {
-    // The raw_val passed to cb might need decoding based on the encoding mask of the "value"
-    // object. We save the mask in decoder and use it to decode the final string that Read should
-    // resolve.
-    PrimeValue decoder;
-    decoder.ImportExternal(value);
-
-    auto cb = [future, decoder = std::move(decoder)](bool is_raw, const string* raw_val) mutable {
-      future.Resolve(DecodeString(is_raw, *raw_val, std::move(decoder)));
-      return false;  // was not modified
-    };
-
-    op_manager_->Enqueue(KeyRef(dbid, key), value.GetExternalSlice(), std::move(cb));
-  }
-  return future;
+  util::fb2::Future<std::string> fut;
+  Read(dbid, key, value, [fut](const std::string& value) mutable { fut.Resolve(value); });
+  return fut;
 }
 
 void TieredStorage::Read(DbIndex dbid, std::string_view key, const PrimeValue& value,
                          std::function<void(const std::string&)> readf) {
-  DCHECK(value.IsExternal());
-  if (value.IsCool()) {
-    util::fb2::Future<string> res = Read(dbid, key, value);
-    readf(res.Get());
-  } else {
-    PrimeValue decoder;
-    decoder.ImportExternal(value);
-
-    auto cb = [readf = std::move(readf), decoder = std::move(decoder)](
-                  bool is_raw, const string* raw_val) mutable {
-      readf(DecodeString(is_raw, *raw_val, std::move(decoder)));
-      return false;
-    };
-    op_manager_->Enqueue(KeyRef(dbid, key), value.GetExternalSlice(), std::move(cb));
-  }
+  PrimeValue decoder;
```

Comment:
please keep dchecks verifying it's external, non cold item

Reply:
true, lost them somehow